### DISCUSSION TITLE
Koka Perceus: Garbage Free Reference Counting with Reuse

### DISCUSSION THREAD

**Discussion by voidexpr** [aardappel/lobster#172] [General]
Is the Perceus approach similar to Lobster ? https://www.microsoft.com/en-us/research/uploads/prod/2020/11/perceus-tr-v1.pdf
maybe it provides some of the formalism behind the compile-time rules to suppress unnecessary reference increment/decrement ? 
It's part of the interesting Koka language: https://koka-lang.github.io/koka/doc/index.html

**Comment by aardappel** [comment]
Was actually just asked this on the `/r/ProgrammingLanguages` discord:

I'm not super familiar with Koka.. it has functional semantics though, so it is likely a much simpler algorithm (looking at the Perceus paper they mostly appear to focus on matching incs with decs for elimination).
Lobster goes a bit further, by specializing functions as part of liveness analysis, allowing things to stay free of runtime refcount ops across function boundaries etc.
